operator/reconciler.go

```diff
@@ -0,0 +1,276 @@
+	RequeueAfter *time.Duration
+	// State can be used alongside RequeueAfter to add the provided state map to the ReconcileRequest supplied in the
+	// future Reconcile call. This allows a Reconcile to "partially complete" and not have to re-do tasks
+	// if it needs to wait on an additional bit of information or if a particular call results in a transient failure.
+	State map[string]any
+}
+
+// Reconciler is an interface which describes an object which implements simple Reconciliation behavior.
+type Reconciler interface {
+	// Reconcile should be called whenever any action is received for a relevant object.
+	// The action and object at the time the action was received are contained within the ReconcileRequest.
+	// If the returned ReconcileResult has a non-nil RequeueAfter, the managing controller should requeue
+	// the Reconcile action, with the same ReconcileRequest and context, after that duration has elapsed.
+	// If the call returns an error, the Reconcile action should be requeued according to the retry policy
+	// of the controller.
+	Reconcile(ctx context.Context, req ReconcileRequest) (ReconcileResult, error)
+}
+
+// ReconcileActionFromResourceAction returns the equivalent ReconcileAction from a provided ResourceAction.
+// If there is no equivalent, it returns ReconcileActionUnknown.
+func ReconcileActionFromResourceAction(action ResourceAction) ReconcileAction {
+	switch action {
+	case ResourceActionCreate:
+		return ReconcileActionCreated
+	case ResourceActionUpdate:
+		return ReconcileActionUpdated
+	case ResourceActionDelete:
+		return ReconcileActionDeleted
+	default:
+		return ReconcileActionUnknown
+	}
+}
+
+// ResourceActionFromReconcileAction returns the equivalent ResourceAction from a provided ReconcileAction.
+// If there is no equivalent, it returns an empty ResourceAction.
+func ResourceActionFromReconcileAction(action ReconcileAction) ResourceAction {
+	switch action {
+	case ReconcileActionCreated:
+		return ResourceActionCreate
+	case ReconcileActionUpdated:
+		return ResourceActionUpdate
+	case ReconcileActionDeleted:
+		return ResourceActionDelete
+	default:
+		return ResourceAction("")
+	}
+}
+
+func NewOpinionatedReconciler(client PatchClient, finalizer string) (*OpinionatedReconciler, error) {
+	if client == nil {
+		return nil, fmt.Errorf("client cannot be nil")
+	}
+	if finalizer == "" {
+		return nil, fmt.Errorf("finalizer cannot be empty")
+	}
+	return &OpinionatedReconciler{
+		finalizer: finalizer,
+		client:    client,
+	}, nil
+}
+
+// OpinionatedReconciler wraps an ordinary Reconciler with finalizer-based logic to convert "Created" events into
+// "resync" events on start-up when the reconciler has handled the "created" event on a previous run,
+// and ensures that "delete" events are not missed during reconciler down-time by using the finalizer.
+type OpinionatedReconciler struct {
+	ReconcileFunc func(context.Context, ReconcileRequest) (ReconcileResult, error)
```

Comment:
Same as in admission control – I suggest wrapping the `Reconciler` interface, rather than adding an extra function type.